File: src/main/python/mmlspark/explainers/LocalExplainer.py
# Copyright (C) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See LICENSE in project root for information.


import sys


if sys.version >= "3":
    basestring = str

from pyspark.ml.param.shared import *
from pyspark.ml.common import inherit_doc
from mmlspark.core.schema.Utils import ComplexParamsMixin
from pyspark.sql import DataFrame
from pyspark.ml.wrapper import JavaParams, JavaWrapper


@inherit_doc
class LocalExplainer(ComplexParamsMixin, JavaParams):
    def __init__(self, java_obj=None) -> None:
        super().__init__()
        self._java_obj = java_obj

    def explain(self, dataframe: DataFrame) -> DataFrame:
        self._transfer_params_to_java()
        return DataFrame(self._java_obj.explain(dataframe._jdf), dataframe.sql_ctx)

    def _new_java_obj(self, java_class, *args):
        return JavaWrapper._new_java_obj(java_class, *args)

File: core/src/main/scala/com/microsoft/azure/synapse/ml/explainers/LocalExplainer.scala
// Copyright (C) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE in project root for information.

package com.microsoft.azure.synapse.ml.explainers

import org.apache.spark.ml.linalg.SQLDataTypes.MatrixType
import org.apache.spark.ml.param.shared.HasOutputCol
import org.apache.spark.ml.{ComplexParamsWritable, Transformer}
import org.apache.spark.sql.DataFrame
import org.apache.spark.sql.types.StructType

trait LocalExplainer
  extends Transformer with HasExplainTarget with HasOutputCol with HasModel with ComplexParamsWritable {

  final def setOutputCol(value: String): this.type = this.set(outputCol, value)

  protected override def validateSchema(inputSchema: StructType): Unit = {
    super.validateSchema(inputSchema)
    if (inputSchema.fieldNames.contains(getOutputCol)) {
      throw new IllegalArgumentException(s"Input schema already has column $getOutputCol")
    }
  }

  override def transformSchema(schema: StructType): StructType = {
    this.validateSchema(schema)
    schema.add(getOutputCol, MatrixType)
  }

  protected def preprocess(df: DataFrame): DataFrame = df
}

object LocalExplainer {
  object LIME {
    def tabular: TabularLIME = {
      new TabularLIME()
    }

    def vector: VectorLIME = {
      new VectorLIME()
    }

    def image: ImageLIME = {
      new ImageLIME()
    }

    def text: TextLIME = {
      new TextLIME()
    }
  }

  object KernelSHAP {
    def tabular: TabularSHAP = {
      new TabularSHAP()
    }

    def vector: VectorSHAP = {
      new VectorSHAP()
    }

    def image: ImageSHAP = {
      new ImageSHAP()
    }

    def text: TextSHAP = {
      new TextSHAP()
    }
  }
}
